pallets/liquidity-mining/src/lib.rs

```diff
@@ -0,0 +1,709 @@
+		),
+
+		/// New liquidity(AMM) pool was added to farm [farm_id, amm_pool_id, liquidity_pool]
+		LiquidityPoolAdded(PoolId, T::AccountId, LiquidityPool<T::BlockNumber, Balance>),
+	}
+
+	#[pallet::storage]
+	#[pallet::getter(fn pool_id)]
+	pub type PoolIdSeq<T: Config> = StorageValue<_, PoolId, ValueQuery>;
+
+	#[pallet::storage]
+	#[pallet::getter(fn global_pool)]
+	pub type GlobalPoolData<T: Config> = StorageMap<
+		_,
+		Twox64Concat,
+		PoolId,
+		GlobalPool<T::BlockNumber, Balance, T::CurrencyId, T::AccountId, T::BlockNumber>,
+		OptionQuery,
+	>;
+
+	#[pallet::storage]
+	#[pallet::getter(fn liquidity_pool)]
+	pub type LiquidityPoolData<T: Config> = StorageDoubleMap<
+		_,
+		Twox64Concat,
+		PoolId,
+		Twox64Concat,
+		T::AccountId, //amm_pool_id
+		LiquidityPool<T::BlockNumber, Balance>,
+		OptionQuery,
+	>;
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Create new liquidity mining program
+		///
+		/// Parameters:
+		/// - `origin`:
+		/// - `total_rewards`:
+		/// - `planned_yielding_periods`: planned number of periods to distribute rewards. WARN: this is not
+		/// how long will farming run.  Owner can destroy farm sooner or liq. mining can run longer
+		/// if all the rewards will not distributed.
+		/// - `blocks_per_period`:
+		/// - `incetivized_token`
+		/// - `reward_currency`
+		/// - `admin_account`
+		/// - `yield_per_period`
+		#[pallet::weight(1000)]
+		#[transactional]
+		pub fn create_farm(
+			origin: OriginFor<T>,
+			total_rewards: Balance,
+			planned_yielding_periods: T::BlockNumber,
+			blocks_per_period: T::BlockNumber,
+			incentivized_token: T::CurrencyId,
+			reward_currency: T::CurrencyId,
+			owner: T::AccountId,
+			yield_per_period: Permill,
+		) -> DispatchResult {
+			T::CreateOrigin::ensure_origin(origin)?;
+
+			Self::validate_create_farm_data(
+				total_rewards,
+				planned_yielding_periods,
+				blocks_per_period,
+				yield_per_period,
+			)?;
+
+			ensure!(
+				T::MultiCurrency::free_balance(reward_currency, &owner) >= total_rewards,
+				Error::<T>::InsufficientRewardCurrencyBalance
+			);
+
+			let planned_periods =
+				Balance::from(TryInto::<u128>::try_into(planned_yielding_periods).map_err(|_e| Error::<T>::Overflow)?);
+			let max_reward_per_period = total_rewards.checked_div(planned_periods).ok_or(Error::<T>::Overflow)?;
+			let now_period = Self::get_now_period(blocks_per_period)?;
+			let pool_id = Self::get_next_id()?;
+
+			let pool = GlobalPool::new(
+				pool_id,
+				now_period,
+				reward_currency,
+				yield_per_period,
+				planned_yielding_periods,
+				blocks_per_period,
+				owner,
+				incentivized_token,
+				max_reward_per_period,
+			);
+
+			<GlobalPoolData<T>>::insert(&pool.id, &pool);
+
+			let pool_account = Self::pool_account_id(pool.id)?;
+			T::MultiCurrency::transfer(reward_currency, &pool.owner, &pool_account, total_rewards)?;
+
+			Self::deposit_event(Event::FarmCreated(pool.id, pool));
```

Comment:
Perhaps additional data would be useful. Consider that if FE processor needs to listen some of the events - what date would it be interested in ?